File: frontend/lib/api/companies.ts
/**
 * lib/api/companies.ts
 */

import { apiRequest } from "./client";

export interface CompanyOut {
  id: number;
  name?: string;
  image?: string;
  created_at?: string;
}

export interface CompanyCreate {
  name: string;
  image?: string;
}

export interface CompanyUpdate {
  name?: string;
  image?: string;
}

export async function getCompanies(skip = 0, limit = 100): Promise<CompanyOut[]> {
  return apiRequest<CompanyOut[]>("/companies", { params: { skip, limit } });
}

export async function getCompany(companyId: number): Promise<CompanyOut> {
  return apiRequest<CompanyOut>(`/companies/${companyId}`);
}

export async function createCompany(data: CompanyCreate): Promise<CompanyOut> {
  return apiRequest<CompanyOut>("/companies", { method: "POST", body: data });
}

export async function updateCompany(companyId: number, data: CompanyUpdate): Promise<CompanyOut> {
  return apiRequest<CompanyOut>(`/companies/${companyId}`, { method: "PATCH", body: data });
}

export async function deleteCompany(companyId: number): Promise<void> {
  return apiRequest<void>(`/companies/${companyId}`, { method: "DELETE" });
}

File: backend/app/routers/companies.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from app.core.database import get_db
from app.models import Company
from app.schemas.auth.company import CompanyCreate, CompanyUpdate, CompanyOut

router = APIRouter(prefix="/companies", tags=["Companies"])


@router.get("/", response_model=list[CompanyOut])
def get_all_companies(skip: int = 0, limit: int = 100, db: Session = Depends(get_db)):
    return db.query(Company).offset(skip).limit(limit).all()


@router.get("/{company_id}", response_model=CompanyOut)
def get_company(company_id: int, db: Session = Depends(get_db)):
    company = db.query(Company).filter(Company.id == company_id).first()
    if not company:
        raise HTTPException(status_code=404, detail="Company not found")
    return company


@router.post("/", response_model=CompanyOut, status_code=201)
def create_company(data: CompanyCreate, db: Session = Depends(get_db)):
    company = Company(**data.model_dump())
    db.add(company)
    db.commit()
    db.refresh(company)
    return company


@router.patch("/{company_id}", response_model=CompanyOut)
def update_company(company_id: int, data: CompanyUpdate, db: Session = Depends(get_db)):
    company = db.query(Company).filter(Company.id == company_id).first()
    if not company:
        raise HTTPException(status_code=404, detail="Company not found")
    for field, value in data.model_dump(exclude_unset=True).items():
        setattr(company, field, value)
    db.commit()
    db.refresh(company)
    return company


@router.delete("/{company_id}", status_code=204)
def delete_company(company_id: int, db: Session = Depends(get_db)):
    company = db.query(Company).filter(Company.id == company_id).first()
    if not company:
        raise HTTPException(status_code=404, detail="Company not found")
    db.delete(company)
    db.commit()